File: src/main/java/io/github/chindeaone/collectiontracker/gui/overlays/AxeAbilityOverlay.java
package io.github.chindeaone.collectiontracker.gui.overlays;

import io.github.chindeaone.collectiontracker.config.ConfigAccess;
import io.github.chindeaone.collectiontracker.config.core.Position;
import io.github.chindeaone.collectiontracker.utils.HypixelUtils;
import io.github.chindeaone.collectiontracker.utils.chat.ChatListener;
import io.github.chindeaone.collectiontracker.utils.rendering.RenderUtils;
import io.github.chindeaone.collectiontracker.utils.rendering.TextUtils;
import io.github.chindeaone.collectiontracker.utils.tab.ForagingStatsWidget;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.Font;
import net.minecraft.client.gui.GuiGraphics;
import net.minecraft.network.chat.Component;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class AxeAbilityOverlay implements AbstractOverlay{

    private final Position position = ConfigAccess.getAxeAbilityPosition();
    private final List<String> axeAbilityOverlayLines = new ArrayList<>();
    private boolean renderingAllowed  = true;
    private boolean expiredTitleShown = true;
    private boolean readyTitleShown = true;

    @Override
    public String overlayLabel() {
        return "Axe Ability";
    }

    @Override
    public Position position() {
        return this.position;
    }

    @Override
    public boolean isEnabled() {
        return ConfigAccess.isAxeAbilityDisplayed() && HypixelUtils.isOnSkyblock();
    }

    @Override
    public boolean isRenderingAllowed() {
        return renderingAllowed;
    }

    @Override
    public void setRenderingAllowed(boolean allowed) {
        this.renderingAllowed = allowed;
    }

    @Override
    public void render(GuiGraphics context) {
        if (!isEnabled()) return;
        List<String> lines = getAxeAbilityLines();

        if (lines.isEmpty()) return;

        RenderUtils.drawOverlayFrame(context, position, () ->
                RenderUtils.renderStrings(context, lines)
        );
    }

    @Override
    public void updateDimensions() {
        if (!isEnabled()) return;
        List<String> lines = getAxeAbilityLines();
        if (lines.isEmpty()) return;

        Font fr = Minecraft.getInstance().font;
        int maxW = 0;
        for (String l : lines) maxW = Math.max(maxW, fr.width(l));
        int h = fr.lineHeight * lines.size();

        position.setDimensions(maxW, h);
    }

    private List<String> getAxeAbilityLines() {
        axeAbilityOverlayLines.clear();

        if (!ConfigAccess.isAxeAbilityDisplayed()) return Collections.emptyList();
        if (ConfigAccess.isAxeAbilityInForagingIslandsOnly() && ForagingStatsWidget.getCurrentForagingIsland() == null) return Collections.emptyList();

        String abilityName = ConfigAccess.getAxeAbilityName();
        String displayName = abilityName.isEmpty() ? "Unknown Ability" : abilityName;

        double cooldown = ChatListener.getFinalAxeCooldown();
        double duration = ChatListener.getFinalAxeDuration();

        if (duration > 0) {
            expiredTitleShown = false;
            readyTitleShown = false;
        }

        if (ConfigAccess.isColeweightAbilityFormat()) {
            // Title logic for Coleweight format
            if (ConfigAccess.isShowAxeAbilityTitle() && duration == 0) {
                if (!expiredTitleShown && cooldown > 0) {
                    String titleExpired = "§6[§3§kd§6] §b§l" + displayName + " §cExpired! §6[§3§kd§6]";
                    RenderUtils.showTitle(Component.literal(titleExpired), ConfigAccess.getTitleDisplayTimer());
                    expiredTitleShown = true;
                }
                if (cooldown == 0) {
                    if (!readyTitleShown) {
                        String titleReady = "§6[§3§kd§6] §b§l" + displayName + " §6[§3§kd§6]";
                        RenderUtils.showTitle(Component.literal(titleReady), ConfigAccess.getTitleDisplayTimer());
                        readyTitleShown = true;
                    }
                } else {
                    readyTitleShown = false;
                }
            }

            // Lines logic for Coleweight format
            String status;
            if (!ConfigAccess.isAbilityCooldownOnly() && duration > 0) {
                status = "§a" + TextUtils.formatTime(duration);
            } else if (cooldown > 0) {
                status = "§c" + TextUtils.formatTime(cooldown);
            } else {
                status = "§aReady!";
            }
            axeAbilityOverlayLines.add("§e" + displayName + " CD: " + status);
        } else {
            // Original Title logic
            if (ConfigAccess.isShowAxeAbilityTitle() && duration == 0) {
                if (!expiredTitleShown && cooldown > 0) {
                    RenderUtils.showTitle(Component.literal("§6" + displayName + " §cExpired!"), ConfigAccess.getTitleDisplayTimer());
                    expiredTitleShown = true;
                }
                if (cooldown == 0) {
                    if (!readyTitleShown) {
                        RenderUtils.showTitle(Component.literal("§6" + displayName + " §aReady!"), ConfigAccess.getTitleDisplayTimer());
                        readyTitleShown = true;
                    }
                } else {
                    readyTitleShown = false;
                }
            }

            // Original format lines
            if (abilityName.isEmpty()) {
                axeAbilityOverlayLines.add("§cUnknown Ability");
            } else {
                axeAbilityOverlayLines.add("§bAxe Ability: §e" + abilityName);
            }

            if (!ConfigAccess.isAbilityCooldownOnly() && duration > 0) {
                axeAbilityOverlayLines.add("§aActive: " + TextUtils.formatTime(duration));
            } else if (cooldown > 0) {
                axeAbilityOverlayLines.add("§cCooldown: " + TextUtils.formatTime(cooldown));
            } else {
                axeAbilityOverlayLines.add("§aReady!");
            }
        }
        return axeAbilityOverlayLines;
    }
}

File: src/main/java/io/github/chindeaone/collectiontracker/gui/overlays/AxeAbilityOverlay.kt
package io.github.chindeaone.collectiontracker.gui.overlays

import io.github.chindeaone.collectiontracker.ModLoader
import io.github.chindeaone.collectiontracker.config.abilityCooldownOnly
import io.github.chindeaone.collectiontracker.config.abilityNameAxe
import io.github.chindeaone.collectiontracker.config.axeAbilityDisplayIndicator
import io.github.chindeaone.collectiontracker.config.axeAbilityInForagingIslandsOnly
import io.github.chindeaone.collectiontracker.config.axeAbilityPosition
import io.github.chindeaone.collectiontracker.config.categories.Misc
import io.github.chindeaone.collectiontracker.config.core.Position
import io.github.chindeaone.collectiontracker.config.displayAxeAbility
import io.github.chindeaone.collectiontracker.config.showAxeExpiredAbilityTitle
import io.github.chindeaone.collectiontracker.config.showAxeReadyAbilityTitle
import io.github.chindeaone.collectiontracker.utils.StringUtils
import io.github.chindeaone.collectiontracker.utils.chat.ChatListener.finalCooldown
import io.github.chindeaone.collectiontracker.utils.chat.ChatListener.finalDuration
import io.github.chindeaone.collectiontracker.utils.render.RenderUtils.renderCooldownBar
import io.github.chindeaone.collectiontracker.utils.render.RenderUtils.renderCooldownCircle
import io.github.chindeaone.collectiontracker.utils.render.RenderUtils.showTitle
import io.github.chindeaone.collectiontracker.utils.world.IslandTracker
import net.minecraft.client.gui.GuiGraphicsExtractor
import net.minecraft.network.chat.Component

class AxeAbilityOverlay : AbstractOverlay() {
    private var cachedLines: List<String> = emptyList()

    private var expiredTitleShown = true
    private var readyTitleShown = true

    override val overlayLabel: String = "Axe Ability"

    override val position: Position get() = axeAbilityPosition

    override val isEnabled: Boolean get() = displayAxeAbility && (!axeAbilityInForagingIslandsOnly || IslandTracker.isForagingIsland())

    override fun render(context: GuiGraphicsExtractor) {
        super.render(context)

        if (!isEnabled) return
        when (axeAbilityDisplayIndicator) {
            Misc.AbilityDisplayIndicator.CROSSHAIR_CIRCLE -> renderCooldownCircle(context, "axe")
            Misc.AbilityDisplayIndicator.CROSSHAIR_BAR -> renderCooldownBar(context, "axe")
            else -> {}
        }
    }

    override fun updateDimensions() {
        if (!isEnabled) return
        updateLinesIfNeeded()

        super.updateDimensions()
    }

    override val lines: List<String>
        get() {
            updateLinesIfNeeded()
            return cachedLines
        }

    private fun updateLinesIfNeeded() {
        if (!isEnabled) {
            cachedLines = emptyList()
            return
        }

        if (ModLoader.clientTicks % 5L != 0L) return

        val abilityName = abilityNameAxe
        val cooldown = finalCooldown
        val active = finalDuration

        val displayName = abilityName.ifEmpty { "Unknown Ability" }

        if (active > 0) {
            expiredTitleShown = false
            readyTitleShown = false
        }

        if (active == 0.0) {
            if (showAxeExpiredAbilityTitle && !expiredTitleShown && cooldown > 0) {
                val titleExpired = "§6[§3§kd§6] §b§l$displayName §cExpired! §6[§3§kd§6]" // Credit to Ninjune for Coleweight's formatting
                showTitle(Component.literal(titleExpired))
                expiredTitleShown = true
            }

            if (cooldown <= 0) {
                if (showAxeReadyAbilityTitle && !readyTitleShown) {
                    val titleReady = "§6[§3§kd§6] §b§l$displayName §6[§3§kd§6]" // Credit to Ninjune for Coleweight's formatting
                    showTitle(Component.literal(titleReady))
                    readyTitleShown = true
                }
            } else {
                readyTitleShown = false
            }
        }

        val status = if (!abilityCooldownOnly && active > 0) {
            "§a" + StringUtils.formatTimeInSeconds(active)
        } else if (cooldown > 0) {
            "§c" + StringUtils.formatTimeInSeconds(cooldown)
        } else {
            "§aReady!"
        }
        cachedLines = listOf("§e$displayName CD: $status") // Credit to Ninjune for Coleweight's formatting
    }
}